services/api/src/models/group.ts

```diff
@@ -922,25 +842,72 @@ export const Group = (clients: {
             // @ts-ignore
             groupId: members[u].roleSubgroupId
           });
+          await purgeGroupInCache(members[u].roleSubgroupId)
         } catch (err) {
           throw new Error(`Could not remove user from group: ${err.message}`);
         }
       }
     }
 
-    // once the users are removed from the group, update the cache
-    const allGroups = await loadAllGroups();
-    const keycloakGroups = await transformKeycloakGroups(allGroups);
-    const data = Buffer.from(JSON.stringify(keycloakGroups)).toString('base64')
+    return await loadGroupById(group.id);
+  };
+
+  const purgeGroupInCache =async (groupId: string) => {
+    // lock the redis interactions
+    let lock = await redlock.acquire(["purgelock"], 5000);
     try {
+      let keycloakGroups = []
+      // attempt to retrieve the data from the cache, or fall back to getting from keycloak
+      try {
+        // check redis for the allgroups cache value
+        const data = await getRedisKeycloakCache("allgroups");
+        let buff = new Buffer(data, 'base64');
+        keycloakGroups = JSON.parse(buff.toString('utf-8'));
+      } catch (err) {
+        lock = await lock.extend(5000);
+        logger.warn(`Couldn't check redis keycloak cache: ${err.message}`);
+        // if it can't be recalled from redis, get the data from keycloak
+        const allGroups = await loadAllGroups();
+        keycloakGroups = await transformKeycloakGroups(allGroups);
+        const data = Buffer.from(JSON.stringify(keycloakGroups)).toString('base64')
+        try {
+          // then attempt to save it to redis
+          await saveRedisKeycloakCache("allgroups", data);
+        } catch (err) {
+          logger.warn(`Couldn't save redis keycloak cache: ${err.message}`);
+        }
+      }
+      let fixedKeycloakGroups = keycloakGroups
+      try {
+        // try and replace or update the group within the cache
+        const group = await loadGroupById(groupId);
+        let exists = false;
+        for (const g of keycloakGroups) {
+          if (g.id == groupId){
+            exists = true
+            fixedKeycloakGroups = keycloakGroups.map(item => item.id === groupId ? group : item)
```

Comment:
There might be more idiomatic ways of doing this for loop stuff (I've got some suggestions if you want to go into it, but not necessary).

Minimally, I'd add here, though, would be a "break" in this `if` statement so that once the group is matched you don't need to spin through the rest of the items. (if `g.id == groupId` and the loop index is 1 out of like 10K, you're doing a lot of extra cycles).

Reply:
Happy if you want to refactor it out, either comment here or directly in the branch. I don't mind

Reply:
Added the break in the loop if the groupid is found for now